### PR TITLE
Add predicate pushdown

### PR DESCRIPTION
**Pull request by mike-luabase**

### CODE REVIEW

**Comment by mike-luabase** [comment]
@samansmink made the updates here. I tried getting rid of the `data` dir from the commit, but wasn't sure the best way to do that.

**Comment by mike-luabase** [comment]
Enhance query performance by filtering data at the metadata level, reducing the amount of data read during scans.

### **Key Changes**

- **Extended `IcebergManifestEntry`:**
  - Added `lower_bounds` and `upper_bounds` maps to store column statistics.
  
- **Utility Function:**
  - Implemented `IcebergUtils::GetFullPath` to resolve file paths accurately.

- **Metadata Retrieval:**
  - Added `GetEntries` template in `IcebergTable` to fetch relevant manifest entries, excluding deleted ones.

- **Predicate Evaluation:**
  - Created `EvaluatePredicateAgainstStatistics` to assess if data files satisfy query predicates based on their statistics.
  - Integrated logging for transparency during evaluation.

- **Scan Expression Modification:**
  - Updated `MakeScanExpression` to filter `data_file_entries` using predicates before scanning.
  - Logged the number of files before and after filtering for debugging.

- **Binding Function Enhancements:**
  - Enhanced `IcebergScanBindReplace` with additional logging and prepared data files based on predicate results.

This implementation optimizes Iceberg table scans by leveraging metadata for early data filtering, significantly improving query efficiency and resource usage.

File: data/iceberg/generated_spec1_0_001/expected_results/last/count.csv [hunk 1]
```diff
@@ -1,2 +0,0 @@
-count
```
Comment (Mytherin) [inline data/iceberg/generated_spec1_0_001/expected_results/last/count.csv:1]: Do we need to delete all of these files?

File: src/include/avro_codegen/iceberg_manifest_entry_partial.hpp [hunk 2]
```diff
@@ -1,124 +1,388 @@
...
+
+/* This code was generated by avrogencpp 1.13.0-SNAPSHOT. Do not edit.*/
+
+#ifndef MANIFEST_ENTRY_HH_2043678367_H
+#define MANIFEST_ENTRY_HH_2043678367_H
 
 
 #include <sstream>
-#include "boost/any.hpp"
+#include <any>
```
Comment (Mytherin) [inline src/include/avro_codegen/iceberg_manifest_entry_partial.hpp:26]: This code is generated - should this be modified?

File: src/iceberg_functions/iceberg_scan.cpp [hunk 3]
```diff
@@ -15,302 +15,870 @@
...
-	auto delete_colref_expr = make_uniq<ColumnRefExpression>("file_path", "iceberg_scan_deletes");
+    auto data_colref_expr = make_uniq<ColumnRefExpression>("filename", "iceberg_scan_data");
+    auto delete_colref_expr = make_uniq<ColumnRefExpression>("file_path", "iceberg_scan_deletes");
 
-	auto data_filename_expr = GetFilenameExpr(std::move(data_colref_expr));
-	auto delete_filename_expr = GetFilenameExpr(std::move(delete_colref_expr));
+    auto data_filename_expr = GetFilenameExpr(std::move(data_colref_expr));
+    auto delete_filename_expr = GetFilenameExpr(std::move(delete_colref_expr));
 
-	return make_uniq<ComparisonExpression>(ExpressionType::COMPARE_NOT_DISTINCT_FROM, std::move(data_filename_expr),
```
Comment (Mytherin) [inline src/iceberg_functions/iceberg_scan.cpp:60]: All these format changes make the code hard to review - can we leave the old format in place?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Some comments from my side. Could you also look at the failing CI?

**Comment by mike-luabase** [comment]
@Mytherin thanks for the review! I cleaned up the issues here: https://github.com/duckdb/duckdb_iceberg/pull/78

will close this one.

**Comment by mike-luabase** [comment]
closing, see above.

**Pull request closed without merging**